### PR TITLE
fix: embed <--> widget communication issue on safari

### PR DESCRIPTION
**Pull request by LetItRock**
- **What kind of change does this PR introduce?**
**Bug fix:** ticket: #1090 

- **What is the current behavior?**
Currently on Safari, it's not possible to open the Embedded Notification Center Widget.

The issue can be seen on the below screenshots.
Chrome:
Screenshot 2022-08-29 at 00 24 30
Safari:
Screenshot 2022-08-29 at 00 24 05

As you can see on Safari `Embed` sends the `INIT_IFRAME` message before the `Widget` gets the opportunity to register the event listener.

- **What is the new behavior (if this is a feature change)?**
This fix solves the issue described above.
The `Embed <--> Widget` communication is reversed and works based on the `postMessage` API.

---

1. When the `Embed` code is running in the browser it creates the `iframe` DOM element for the `Widget`.
2. Right after it registers the `message` event listener on that `iframe` to receive events from the `Widget`.
3. When the `Widget` is loaded and React triggers the `useEffect` hook (on mount) - it registers it's own event listener on the `window` object to receive events from the `Embed`.
4. Then right after it sends the `WIDGET_READY` event to the `Embed`.
5. `Embed` receives that event and sends back `INIT_IFRAME` to `Widget` with a configuration object.
6. `Widget` receive it and does the initialization. 

In short:
1. `WIDGET --> WIDGET_READY --> EMBED`
2. `EMBED --> INIT_IFRAME --> WIDGET`
3. Widget initialized.

**Unfortunately, currently it's not possible to run Cypress tests in Safari.
And I haven't checked the fix on the iOS Safari.**

---

- **Screenshots**:

Screenshot 2022-09-03 at 14 00 12
Screenshot 2022-09-03 at 13 59 44
Screenshot 2022-09-03 at 13 58 24
Screenshot 2022-09-03 at 14 02 10

Cypress tests:
Screenshot 2022-09-03 at 23 03 42

### CODE REVIEW

File: apps/widget/src/components/notification-center/NotificationCenterWidget.tsx [hunk 1]
```diff
@@ -33,24 +33,24 @@ export function NotificationCenterWidget(props: INotificationCenterWidgetProps)
 
   useEffect(() => {
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const handler = async (event: { data: any }) => {
-      if (event.data.type === 'INIT_IFRAME') {
-        setUserDataPayload(event.data.value.data);
+    const handler = ({ data }: any) => {
+      if (data && data.type === 'INIT_IFRAME') {
```
Comment (LetItRock) [inline apps/widget/src/components/notification-center/NotificationCenterWidget.tsx:37]: we do only care here about the data object

File: apps/widget/src/components/notification-center/NotificationCenterWidget.tsx [hunk 2]
```diff
@@ -64,6 +64,8 @@ export function NotificationCenterWidget(props: INotificationCenterWidgetProps)
 
     window.addEventListener('message', handler);
 
+    window.parent.postMessage({ type: 'WIDGET_READY' }, '*');
```
Comment (LetItRock) [inline apps/widget/src/components/notification-center/NotificationCenterWidget.tsx:67]: when the `Widget` is mounted - it first registers the event listener to the `message` event and then sends the event to `Embed` saying that it's ready to be initialized. The `window.parent` prop is a reference to the parent window or to self.

File: libs/embed/src/embed.ts [hunk 3]
```diff
@@ -123,13 +123,9 @@ class Novu {
...
-    window.addEventListener('resize', () => {
-      positionIframe();
-    });
-
-    window.addEventListener('click', (e: any) => {
-      if (document.querySelector(this.selector)?.contains(e.target)) {
-        this.widgetVisible = !this.widgetVisible;
+    function handleClick(e: MouseEvent | TouchEvent) {
+      if (document.querySelector(_scope.selector)?.contains(e.target as Node)) {
+        _scope.widgetVisible = !_scope.widgetVisible;
```
Comment (LetItRock) [inline libs/embed/src/embed.ts:128]: `_scope` is `this` in the above code

File: libs/embed/src/embed.ts [hunk 4]
```diff
@@ -196,6 +196,32 @@ class Novu {
   initializeIframe = (clientId: string, options: any) => {
     if (!document.getElementById(IFRAME_ID)) {
       const iframe = document.createElement('iframe');
+      window.addEventListener(
```
Comment (LetItRock) [inline libs/embed/src/embed.ts:199]: Add the `message` event listener right after the `iframe` DOM element is created. This way the `Embed` will never miss the events. Before we were doing it in the `onload` which is too late for some browsers.

File: libs/embed/src/embed.ts [hunk 5]
```diff
@@ -148,7 +144,11 @@ class Novu {
       } else {
         hideWidget();
       }
-    });
+    }
+
+    window.addEventListener('resize', positionIframe);
+    window.addEventListener('click', handleClick);
+    window.addEventListener('touchstart', handleClick);
```
Comment (LetItRock) [inline libs/embed/src/embed.ts:151]: Use the same listener to `click and touchstart` events. This fixes the issue on the touch devices that it doesn't react to the touch events.

File: libs/embed/src/embed.ts [hunk 6]
```diff
@@ -253,29 +279,12 @@ class Novu {
                 }
               }
             },
-            enablePublicMethods: true, // Enable methods within iframe hosted page
```
Comment (LetItRock) [inline libs/embed/src/embed.ts:256]: this prop was removed

**Review by scopsy — APPROVED** [review]
I really love the flow right now, I think it makes a lot of sense here! As usually, a pleasure reviewing your PR. 

@djabarovgeorge could you please take a another look at this and maybe perhaps do a small test locally for it? Just to make sure/

**Comment by djabarovgeorge** [comment]
Hey @LetItRock if there is a way to share the snipped of the iframe host you used? I am a bit struggling to host the embed 😅

**Comment by LetItRock** [comment]
> Hey @LetItRock if there is a way to share the snipped of the iframe host you used? I am a bit struggling to host the embed 😅 

Hey, sure I will explain/send you my setup a little later today ;)

**Comment by LetItRock** [comment]
> Hey @LetItRock if there is a way to share the snipped of the iframe host you used? I am a bit struggling to host the embed 😅

hey, so the steps that I did were these:
1. Run the `API` with `npm run start:api` from the root.
2. Run the `WS` with `npm run start:ws` from the root.
3. Go to the `apps/widget` in the `package.json` change the `start:dev` script PORT to `3500` (this is the expected port from the Embed).
4. From the same folder run `npm run start:dev` - this will be running the `Widget` in dev mode.
5. Then we need to build the embed, unfortunately. Go to the `libs/embed` folder and run the `npm run build` - it will build the `Embed` with the `local` settings.
6. Copy the `embed.umd.min.js` from the `libs/embed/dist` folder.
7. Paste it in the `apps/widget/cypress/test-shell` folder.
8. In the same folder replace the `script` tag at the bottom of the `index.html` with below, but remember to update the  `novu.init` config with your values:
```
<script>
  (function (n, o, t, i, f) {
    n[i] = {};
    var m = ['init'];
    n[i]._c = [];
    m.forEach(
      (me) =>
        (n[i][me] = function () {
          n[i]._c.push([me, arguments]);
        })
    );
    var elt = o.createElement(f);
    elt.type = 'text/javascript';
    elt.async = true;
    elt.src = t;
    var before = o.getElementsByTagName(f)[0];
    before.parentNode.insertBefore(elt, before);
  })(window, document, './embed.umd.min.js', 'novu', 'script');

  novu.init(
    'pO6TprcWFmW9',
    {
      unseenBadgeSelector: '#unseen-badge',
      bellSelector: '#notification-bell',
    },
    {
      subscriberId: '62e00ab53c36d8591d8fb77c',
      email: 'pawel.tymczuk@gmail.com',
      first_name: 'Paweł',
      last_name: 'Tymczuk',
    }
  );
</script>
```
9. Then run the `index.html` with the `http-server` or `VS Code - Live Server Extension`.

The app should look similar to this:
Screenshot 2022-09-04 at 20 48 07

I'm not sure if there is a better way to do this, but I've invented it myself :D

**Comment by djabarovgeorge** [comment]
Hey @LetItRock Thanks for the well-explained guide took me seconds to set it up! , in the past, I did something similar but I used the CLI demo dashboard (which is pretty the same), we most definitely need to create a more friendly way to debug the embed.

@scopsy tested, work like a charm.

@LetItRock Awesome work fixing the issue I really struggled figuring out the source of it 😁
